### PR TITLE
Update of Multi_critical_filtration

### PR DESCRIPTION
**Pull request by hschreiber**
Adapted it to the new version of One_critical_filtration and corrected a few things which didn't work or could be optimized + unit tests

TODO: Still misses proper doc

### CODE REVIEW

File: azure-pipelines.yml
Comment (DavidLapous) [inline azure-pipelines.yml:1]: Whats all this ?

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 1]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
-  Multi_critical_filtration() : multi_filtration_(1) {
-    multi_filtration_[0] = co ? One_critical_filtration<T>::inf() : One_critical_filtration<T>::minus_inf();
-  };  // initialize at 1, with 0 parameter (consistent with below)
-  Multi_critical_filtration(int n)
-      : multi_filtration_({One_critical_filtration<T>(n)}) {};  // minus infinity by default
-  Multi_critical_filtration(int n, T value)
-      : multi_filtration_({One_critical_filtration<T>(n, value)}) {};  // std::vector<T>(n, value){};
-  Multi_critical_filtration(std::initializer_list<T> init)
-      : multi_filtration_({One_critical_filtration<T>(init)}) {};  // : std::vector<T>({std::vector<T>(init)};
+  using Single_point = One_critical_filtration<T>;  // Type of the One critical subtype
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:46]: I'm not a big fan of `Single_point`. `generator` ?

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 2]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
+  {
+    std::vector<std::vector<T>> out(f.num_generators(), std::vector<T>(f.num_parameters()));
+    const auto &cont = f.get_underlying_container();
+    for (std::size_t i = 0; i < f.num_generators(); ++i) {
+      for (std::size_t j = 0; j < f.num_parameters(); ++j) {
+        out[i][j] = cont[i][j];
+      }
     }
-    return true;
+    return out;
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h]: Same as the one parameter case, conversions should be handled by a proper pxd file

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 3]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
   Multi_critical_filtration(typename std::vector<T>::const_iterator it_begin,
                             typename std::vector<T>::const_iterator it_end)
-      : multi_filtration_({One_critical_filtration<T>(it_begin, it_end)}) {};  // : std::vector<T>(it_begin, it_end){};
+      : multi_filtration_({Single_point(it_begin, it_end)}) {};
   Multi_critical_filtration(const Multi_critical_filtration<T> &other) : multi_filtration_(other.multi_filtration_) {};
 
-  Multi_critical_filtration &operator=(const Multi_critical_filtration<T> &other) {
-    this->multi_filtration_ = other.multi_filtration_;
+  Multi_critical_filtration &operator=(const Multi_critical_filtration<T> &other)
+  {
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:78]: Is the `{` here instead of above, a gudhi convention ?

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 4]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
-    } else {
-      U projection = std::numeric_limits<U>::max();
-      for (auto &y : *this) {  // cannot be const (Gudhi)
-        projection = std::min(projection, y.linear_projection(x));
-      }
-      return projection;
+
+  /*
+   * Adds a birth point to the list of births,
+   * if it is useful (according to the method `dominate`)
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:280]:
```suggestion
   * if it is useful (according to the method `_contains`)
```

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 5]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
    */
-  static inline bool dominates(const OneCritical &a, const OneCritical &b, value_type max_error) {
-    if constexpr (co)
-      return a - max_error <= b;
-    else {
-      return a + max_error >= b;
+  template <typename oned_array>
+  void project_onto_grid(const std::vector<oned_array> &grid, bool coordinate = true)
+  {
+    // not sure why this is needed? Are the tests with the numbers of parameters not enough?
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:312]: I prefer doing asserts rather than long comments, this is more readable

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 6]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
+    if (!coordinate) simplify();
   }
 
-  static inline bool dominates(const OneCritical &a, const OneCritical &b) {
-    if constexpr (co)
-      return a <= b;
-    else {
-      return a >= b;
-    }
+  // no NaN?
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:323]: If there is a nan, I'm not sure what's the best way to handle this

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 7]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
-    Multi_critical_filtration<U> out(this->num_generators());
-    for (std::size_t i = 0; i < this->num_generators(); ++i) {
-      out[i] = this->operator[](i).evaluate_in_grid(grid);
+  // FONCTIONNALITIES
+
+  friend Single_point factorize_below(const Multi_critical_filtration &f)
+  {
+    if (f.num_generators() == 0) return Single_point();
+    Single_point result(f.num_parameters(), Single_point::T_inf);
+    for (const auto &fil : f) {
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h]: morally `f` is the ``filtration value'', not `fil`, but this doens't really matter

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 8]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
+                                                                   const std::vector<std::vector<U>> &grid)
+  {
+    Multi_critical_filtration<U> out;
+    out.set_num_generators(f.num_generators());
+    for (std::size_t i = 0; i < f.num_generators(); ++i) {
+      out[i] = evaluate_coordinates_in_grid(f[i], grid);
     }
-    std::swap(multi_filtration_, out);
+    out.simplify();
+    return out;
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:431]: Isn't this slower than my implementation with add point ?

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h [hunk 9]
```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
...
-    std::vector<std::vector<T>> out(this->num_generators(), std::vector<T>(this->num_parameters()));
-    for (std::size_t i = 0; i < this->num_generators(); ++i) {
-      for (std::size_t j = 0; j < this->num_parameters(); ++j) {
-        out[i][j] = multi_filtration_[i][j];
+
+  // 0 == equal
+  // 1 == a dom b
+  // 2 == b dom a
+  // 3 == none
+  static int _get_domination_relation(const Single_point &a, const Single_point &b)
```
Comment (DavidLapous) [inline src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:494]: This is a bit error-prone / hard to follow. Maybe an enum is better ? I don't know if there is a cost there.

File: src/Persistence_matrix/include/gudhi/persistence_interval.h
Comment (DavidLapous) [inline src/Persistence_matrix/include/gudhi/persistence_interval.h:1]: Is there a link between this and the multipers PR?

File: azure-pipelines.yml
Comment (hschreiber) [reply azure-pipelines.yml:1]: That is just a github bug... I pulled the newest version of Gudhi, the same you did for [Multiparameter-Simplextree-C++-part](https://github.com/DavidLapous/gudhi-devel-multi/tree/Multiparameter-Simplextree-C%2B%2B-part), but for some reasons github still makes the diff on the version from before... So, he sees a difference where is none.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 1 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:46]: Me neither, but I needed something else than `Base` and it should also work for `One_critical_filtration`. `Generator` sounds fine if you don't think it could be misunderstood with some other notion of generator in this context.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 2 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h]: Ok, I remove the method then.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 3 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:78]: No, it is a my own convention and it is automatically applied, so it is like that for all methods. It was already the case in the PR from before. I could undo it, but I find that methods with lot of parameters easier to read that way.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 4 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:280]: I will redo all the documentation in other PR.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 5 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:312]: That was not my question. I was wondering why you needed to assure that `grid.size() >= num_generators()`. This condition does not seem important for the functionality of the method.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 7 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h]: Well, it is a filtration value in a filtration value in some sense? I could change it to `g` for `generator` if you prefer?

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 8 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:431]: I am pretty sure that `evaluate_coordinates_in_grid` didn't used add point before neither? I am not sure what you mean? The only thing I changed here is adding the simplified method. And using `set_num_generators` because `Multi_critical_filtration<U> out(this->num_generators());` is not doing what you want here.

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 9 above)
Comment (hschreiber) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:494]: It is a private method, so security doesn't really matter as long as you know what you are doing with it. Simple `enum`s are no different than `int`s and are not more secure. `enum class` would be the secure version, but it seems a bit overkill as no one outside of this class will ever see them?

File: src/Persistence_matrix/include/gudhi/persistence_interval.h
Comment (hschreiber) [reply src/Persistence_matrix/include/gudhi/persistence_interval.h:1]: No, same problem than for the first comment...

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 5 above)
Comment (DavidLapous) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:312]: yes this is `num_parameters`, oups

File: src/Multi_filtration/include/gudhi/Multi_critical_filtration.h (hunk 8 above)
Comment (DavidLapous) [reply src/Multi_filtration/include/gudhi/Multi_critical_filtration.h:431]: ah ok

**Review by DavidLapous — APPROVED** [review]
